README
PetVM

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.


INSTALLATION

To install this module, run the following commands:

	perl Build.PL
	./Build
	./Build test
	./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc PetVM

You can also look for information at:

    RT, CPAN's request tracker (report bugs here)
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=PetVM

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/PetVM

    CPAN Ratings
        http://cpanratings.perl.org/d/PetVM

    Search CPAN
        http://search.cpan.org/dist/PetVM/


LICENSE AND COPYRIGHT

Copyright (C) 2012 Dmitriy Ryajov

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for PetVM

0.01    Date/time
        First version, released on an unsuspecting world.



test.pl
#!perl

use strict;
use warnings;

use PetVM::PetParser;
use IO::Scalar;
use Data::Dumper;

# buffer to hold the parsed syntax
my $buf;

# test PUSH numeric
$buf = << 'EOF';

// init counter
PUSH    1000
PUSH    0
STORE           // store away the initial counter

// loop untill 0
LOOP:
	PUSH    1
	SUBST                               // n - 1
	PUSH    0
	STORE                               // put result in address 0
	PUSH   "\n"
	PUSH    0                           
	LOAD                                // load value from address 0 on to the stack
	PUSH    "VALUE IS: "                // put string on stack
	OUT
	OUT
	OUT
	PUSH    0                           
	LOAD                                // load value from address 0 on to the stack
	PUSH    0                           // push value to compare to
JMPNEQ   [LOOP]

EOF

# create a new parser
my $parser = PetVM::PetParser->new;

$parser->parse( IO::Scalar->new( \$buf ) );

my $instructions = $parser->get_opcodes();

my $pet = PetVM->new( instructions => $instructions );
$pet->run();

#print Dumper($pet);


lib/PetVM/PetLexer.pm
package PetVM::PetLexer;

use 5.006;
use strict;
use warnings;

=head1 NAME

PetVM::PetLexer - The great new PetVM::PetLexer!

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';

use Lexer::Lex;

use strict;

sub _print {
	my $lex = shift;
	print "LABEL: ", $lex->last()->label, " - ";
	print $lex->last()->content, "\n";

	return;
}

sub new {
	my $class = shift;
	my $self  = {};

	bless $self, $class;

	my $lex = $self->init_lexer();
	$self->{lexer} = $lex;

	return $self;
}

sub lexer {
	my $self = shift;

	return $self->{lexer};
}

sub init_lexer {

	# OPCODES
	my $opcodes =  qr/ADD|SUBST|DIV|MUL|MOD|SHL|SHR|AND|OR|
	                   COMP|XOR|POP|CALL|RET|OUT|IN|STORE|LOAD/xo;
	                   
    my $jmp_opcodes = qr/JMPEQ|JMPNEQ|JMPGT|JMPLT|JMP/xo;

	# Match strings
	my $string = qr{
                  ".*"  # match anything within ""
               }xom;

	# Match numbers
	my $number = qr{
                  \d+  # match a number
               }xom;

	# match comments
	my $comments = qr{//*.+}ox;

    # these are ignored
    my $ignored = qr{^[\s|\n]+?}ox;

	my @tokens = (
		[
		  'OP',
		  qr/$opcodes+/x,
		],
        [
          'OP_PUSH',
          qr/PUSH/x,
        ],
        [
          'OP_JMP',
          qr/$jmp_opcodes/x,
        ],
		[
		  'STRING',
		  qr/$string/x,
          sub {
            my $lex = shift;
            my $content = $lex->last()->content;
            $content =~ s/\"//g;
            $lex->last()->content($content);
          } 		  
		],
		[
		  'NUMBER',
		  qr/$number/x,
		],
		[
		  'LABEL',
		  qr/\w+:/x,
          sub {
            my $lex = shift;
            my $content = $lex->last()->content;
            $content =~ s/:$//g;
            $lex->last()->content($content);
          }		  
		],
		[
		  'LABEL_REF',
		  qr/\[\w+\]/x,
		  sub {
		  	my $lex = shift;
		  	my $content = $lex->last()->content;
		  	$content =~ s/\[(.*)\]/$1/g; # strip [], we don't need that
		  	$lex->last()->content($content);
		  }
		],
		[ 'IGNORE', qr/$ignored/x, undef, 1 ],
		[ 'COMMENTS', qr/$comments/x, undef, 1 ],
	);

	my $lexer = Lexer::Lex->new(
		\@tokens,
		sub {
			my $lex = shift;
			my $tok = $lex->{_unmatch};
			print $tok->label,   "\n";
			print $tok->content, "\n";
			die "Unable to parse line, " . $lex->get_line . "!\n";
		},
		"\n"
	);

	return $lexer;
}

sub token {
	my $self = shift;

    my $token = $self->lexer->token;
    my ($label, $content) = ( $token->label, $token->content ); 

    if ($label eq 'EOF') {
    	($label, $content) = ('', undef);
    } elsif ($label eq 'NEWLINE') {
    	($label) = '';
    }

	return ($label, $content);
}

sub from {
	my ( $self, $file ) = @_;
	$self->lexer->from($file);
	return;
}

1;

__END__

=head1 SYNOPSIS

Quick summary of what the module does.

Perhaps a little code snippet.

    use PetVM::PetLexer;

    my $foo = PetVM::PetLexer->new();
    ...

=head1 EXPORT

A list of functions that can be exported.  You can delete this section
if you don' t export anything, such as for a purely object-oriented module .

=head1 SUBROUTINES/METHODS

=head2 function1

=cut

			 sub function1 {
		   }

=head2 function2

=cut

		   sub function2 {
		   }

=head1 AUTHOR

Dmitriy Ryajov, C<< <dryajov at gmail.com> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-petvm at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=PetVM>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.




=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc PetVM::PetLexer


You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=PetVM>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/PetVM>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/PetVM>

=item * Search CPAN

L<http://search.cpan.org/dist/PetVM/>

=back


=head1 ACKNOWLEDGEMENTS


=head1 LICENSE AND COPYRIGHT

Copyright 2012 Dmitriy Ryajov.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.


=cut

		   1;    # End of PetVM::PetLexer


blib/lib/PetVM/PetParser.pm
####################################################################
#
#    This file was generated using Parse::Yapp version 1.05.
#
#        Don't edit this file, use source file instead.
#
#             ANY CHANGE MADE HERE WILL BE LOST !
#
####################################################################
package PetVM::PetParser;
use vars qw ( @ISA );
use strict;

@ISA= qw ( Parse::Yapp::Driver );
use Parse::Yapp::Driver;

#line 1 "PetGrammar.yp"

use PetVM qw(%opcodes);
use PetVM::PetLexer;
use Data::Dumper;


sub new {
        my($class)=shift;
        ref($class)
    and $class=ref($class);

    my($self)=$class->SUPER::new( yyversion => '1.05',
                                  yystates =>
[
	{#State 0
		ACTIONS => {
			'' => -1,
			'LABEL' => -1,
			'OP' => -1,
			'error' => 1,
			'OP_JMP' => -1,
			'OP_PUSH' => -1
		},
		GOTOS => {
			'program' => 2
		}
	},
	{#State 1
		DEFAULT => -3
	},
	{#State 2
		ACTIONS => {
			'' => 3,
			'OP' => 6,
			'LABEL' => 5,
			'OP_JMP' => 7,
			'OP_PUSH' => 8
		},
		GOTOS => {
			'expr' => 4
		}
	},
	{#State 3
		DEFAULT => 0
	},
	{#State 4
		DEFAULT => -2
	},
	{#State 5
		DEFAULT => -8
	},
	{#State 6
		DEFAULT => -4
	},
	{#State 7
		ACTIONS => {
			'LABEL_REF' => 9
		}
	},
	{#State 8
		ACTIONS => {
			'NUMBER' => 11,
			'STRING' => 12
		},
		GOTOS => {
			'arg_num' => 10,
			'arg_string' => 13
		}
	},
	{#State 9
		DEFAULT => -7
	},
	{#State 10
		DEFAULT => -5
	},
	{#State 11
		DEFAULT => -9
	},
	{#State 12
		DEFAULT => -10
	},
	{#State 13
		DEFAULT => -6
	}
],
                                  yyrules  =>
[
	[#Rule 0
		 '$start', 2, undef
	],
	[#Rule 1
		 'program', 0, undef
	],
	[#Rule 2
		 'program', 2, undef
	],
	[#Rule 3
		 'program', 1,
sub
#line 11 "PetGrammar.yp"
{ $_[0]->YYErrok }
	],
	[#Rule 4
		 'expr', 1,
sub
#line 14 "PetGrammar.yp"
{
                                    push @{$_[0]->{YYDATA}->{opcodes}}, 
                                    [$opcodes{$_[0]->get_opcode_from_keyword($_[1])}];
                                    
                                    return $_[1];
                                }
	],
	[#Rule 5
		 'expr', 2,
sub
#line 20 "PetGrammar.yp"
{
                                    push @{$_[0]->{YYDATA}->{opcodes}}, 
                                    [$opcodes{$_[0]->get_opcode_from_keyword($_[1])}, [$_[2]]];
                                    
                                    return ($_[1], $_[2]);
                                }
	],
	[#Rule 6
		 'expr', 2,
sub
#line 26 "PetGrammar.yp"
{
                                    push @{$_[0]->{YYDATA}->{opcodes}}, 
                                    [$opcodes{$_[0]->get_opcode_from_keyword($_[1])}, ["$_[2]"]];
                                    
                                    return ($_[1], $_[2]);
                                }
	],
	[#Rule 7
		 'expr', 2,
sub
#line 32 "PetGrammar.yp"
{
                                    # verify that there is a valid jump point
                                    exists $_[0]->{YYDATA}->{labels}->{$_[2]} or do {
                                        $_[0]->YYData->{ERRMSG} = "Invalid jump point!\n";
                                        $_[0]->YYError;
                                        return undef;
                                    };
                                    
                                    # get the jump point
                                    my $jmp_pt = $_[0]->{YYDATA}->{labels}->{$_[2]};
                                    
                                    # push it on the opcodes stack
                                    push @{$_[0]->{YYDATA}->{opcodes}}, [$opcodes{'OP_PUSH'}, [$jmp_pt]];
                                    
                                    # then we push the jump instruction on the stack as well
                                    push @{$_[0]->{YYDATA}->{opcodes}}, [$opcodes{$_[0]->get_opcode_from_keyword($_[1])}];
                                    
                                    return ($_[1], $_[2]);
                                }
	],
	[#Rule 8
		 'expr', 1,
sub
#line 51 "PetGrammar.yp"
{
                                    # make the label point one position beyond 
                                    # in the instruction array last instruction.
                                    # we need this because the next instruction 
                                    # is going to be the jump to location
                                    $_[0]->{YYDATA}->{labels}->{$_[1]} = $#{$_[0]->{YYDATA}->{opcodes}}+1;
                                    
                                    return ($_[1]);                            
                                }
	],
	[#Rule 9
		 'arg_num', 1, undef
	],
	[#Rule 10
		 'arg_string', 1, undef
	]
],
                                  @_);
    bless($self,$class);
}

#line 69 "PetGrammar.yp"


sub get_opcode_from_keyword {
    my $self    = shift;
    my $keyword = shift;
    
    return "OP_$keyword";
}

sub _error {

    exists $_[0]->YYData->{ERRMSG}
    and do {
        print $_[0]->YYData->{ERRMSG}, "\n";
        return $_[0]->YYData->{ERRMSG};
    };
        
    die "Syntax error.\n";
}

sub _lexer {
    my ($self) = @_;
    
    my ($label, $value) = $self->{YYDATA}->{lexer}->token;
    
    return ($label, $value);
}

sub init {
    my $self = shift;

    # create the lexer
    my $lexer = new PetVM::PetLexer;
    $lexer->from(@_);
    $self->{YYDATA}->{lexer} = $lexer;
    
    $self->{YYDATA}->{opcodes}  = []; # opcodes array
    $self->{YYDATA}->{labels}   = {}; # labels hash - holds references to opcodes    
    
    return;
}

sub parse {
    my $self = shift;
    my $file = shift;
    
    $self->init($file);
    
    $self->YYParse( yylex => \&_lexer, yyerror => \&_error, yydebug => 0x0 );
    
    return;
}

sub get_opcodes {
    my $self = shift;
    
    return $self->{YYDATA}->{opcodes};
}

1;


lib/PetVM.pm
package PetVM;

=pod

=head1 NAME

PetVM - The Perl experimental and toy VM

=head1 SYNOPSIS

  my $object = PetVM->new(
  );
  $object->dummy;

=head1 DESCRIPTION

The author was too lazy to write a description.

=head1 METHODS

=cut

use 5.010;
use strict;
use warnings;

use Moose;
use Modern::Perl;
use TryCatch;
use IO::Handle;
use Carp;
use Data::Dumper;
use Scalar::Util;

our $VERSION = '0.01';

=pod

=head1 OPCODES

ADD     0
SUBST   1
DIV     2
MUL     3
MOD     4

SHL     5
SHR     6
AND     7
OR      8
COMP    9
XOR     10

JMP     11
JMPEQ   12
JMPNEQ  13
JMPGT   14
JMPLT   15

PUSH    16
POP     17

CALL    18
RET     19

OUT     20
IN      21

STORE   22
LOAD    23
CLS     24

=cut

use Exporter;
use base qw/Exporter/;

our @EXPORT_OK;
our %EXPORT_TAGS;

# setup the instructions hash so that it can be
# imported from parsers and such
BEGIN {

	our %opcodes;    # holds a map of opcode name to opcode id

	@EXPORT_OK = qw(%opcodes)
	  ; # exported by default, all parsers should use this to create the executable

	%EXPORT_TAGS = (
		OPCODES => [
			qw(OP_ADD
			  OP_SUBST
			  OP_DIV
			  OP_MUL
			  OP_MOD
			  OP_SHL
			  OP_SHR
			  OP_AND
			  OP_OR
			  OP_COMP
			  OP_XOR
			  OP_JMP
			  OP_JMPEQ
			  OP_JMPNEQ
			  OP_JMPGT
			  OP_JMPLT
			  OP_PUSH
			  OP_POP
			  OP_CALL
			  OP_RET
			  OP_OUT
			  OP_IN
			  OP_STORE
			  OP_LOAD
			  OP_CLS )
		],
	);

	# export the opcodes
	map { push @EXPORT_OK, $_; } @{ $EXPORT_TAGS{OPCODES} };

	no strict 'refs';
	no strict 'subs';

	# I whish perl had real enums...
	my $const_cnt = 0;
	map {
		*{$_} = sub { $const_cnt++ };    # create constants using subs
		$opcodes{$_} = &{$_};            # map into the opcode hash
	} @{ $EXPORT_TAGS{OPCODES} };
}

=pod

=head2 new

  my $object = PetVM->new();

The C<new> constructor lets you create a new B<PetVM> object.

So no big surprises there...

Returns a new B<PetVM> or dies on error.

=cut

# array of instructions
# each entry contains a reference to another array
# in the form of [[OP, [p1, p2, ... pN]]]
#
# where OP is the opcode code and the array ref is an
# (optional) array of parameters passed to the instruction
has 'instructions' => (
						isa      => 'ArrayRef[ArrayRef]',
						is       => 'rw',
						required => 1
);

# current instruction pointer
# (offset into instructions)
has 'ip' => (
			  isa      => 'Num',
			  is       => 'rw',
			  default  => 0,
			  init_arg => undef,
);

# the execution stack
has 'stack' => (
				 isa      => 'ArrayRef',
				 is       => 'rw',
				 default  => sub { [] },
				 init_arg => undef,
);

# speciall stack used for subrutine
# call return addresses
has 'address_stack' => (
						 isa      => 'ArrayRef',
						 is       => 'rw',
						 default  => sub { [] },
						 init_arg => undef
);

# a pseudo heap
has 'heap' => (
				isa      => 'ArrayRef[ArrayRef]',
				is       => 'rw',
				builder  => '_init_heap',
				init_arg => undef,
);

# build the heap
sub _init_heap {
	my @heap;

	# construct a heap of 256 elements
	for ( 0 .. 255 ) {
		push @heap, [
					  0,        # the used free bit
					  undef,    # the actual value
		];
	}

	return \@heap;
}

# a map of OP codes to internal subrutines
has 'routines' => (
					isa      => 'ArrayRef[CodeRef]',
					is       => 'ro',
					builder  => '_build_routines',
					init_arg => undef,
);

# build the rutines
sub _build_routines {
	return [
			 \&ADD,   \&SUBST, \&DIV,   \&MUL,    \&MOD,
			 \&SHL,   \&SHR,   \&AND,   \&OR,     \&COMP,
			 \&XOR,   \&JMP,   \&JMPEQ, \&JMPNEQ, \&JMPGT,
			 \&JMPLT, \&PUSH,  \&POP,   \&CALL,   \&RET,
			 \&OUT,   \&IN,    \&STORE, \&LOAD,   \&CLS,
	];
}

=pod

=head2 eval

This method does something... apparently.
    
=cut

sub move_next {
	my $self = shift;

	#  return an instruction and increment
	my $ip = $self->instructions->[ $self->ip ];
	$self->ip( $self->ip + 1 );
	return $ip;
}

sub run {
	my $self = shift;

	while ( my $OP = $self->move_next ) {
		$self->op_eval($OP);
	}

	return;
}

sub op_eval {
	my $self = shift;
	my $OP   = shift;

	my $routine = $self->routines->[ $OP->[0] ];
	my $res = $routine->( $self, $OP->[1] );

	return $res;
}

sub _find_free_block {
	my $self = shift;
	my $size = shift;        # size if the block of address to search for
	my $addr = shift || 0;

	my $curr_size = 0;
	foreach my $index ( $addr .. @{ $self->heap } ) {
		if ( $self->heap->[$index]->[0] == 1 ) {
			return $self->_find_free_block( $size, ++$index );
		}
	} continue {
		$index++;
		last
		  if ( ++$curr_size >= $size );
	}

	return ( $addr, $curr_size );
}

sub _peek {
	my $self = shift;

	return $self->stack->[-1];
}

sub _dump_stack {
	my $self = shift;

	my $stack = $self->stack;

	print "DUMPING STACK: ", Dumper($stack), "\n";

	return;
}

sub _trace {
	my $self = shift;
	my $msg  = shift;

	# print a trace message
	carp $msg;

	# dump the stack
	$self->_dump_stack();

	croak "General error, bailing out!!!";
}

sub _is_numeric_pair {
	my $self = shift;

	my ( $a, $b ) = @_;

	return (    Scalar::Util::looks_like_number($a)
			 && Scalar::Util::looks_like_number($b) );
}

# IMPLEMENT THE OPCODE ROUTINES

# STACK MANIPULATION

# pushes an element onto the stack
sub PUSH {
	my $self = shift;
	my $arg  = shift;

	push @{ $self->stack }, ref $arg ? $arg->[0] : $arg;

	#print Dumper( \@{ $self->stack } );
	return;
}

# pops an element from the stack
sub POP {
	my $self = shift;

	my $val = pop @{ $self->stack };

	#print Dumper( \@{ $self->stack } );

	return $val;
}

# ARITHMETIC

# adds two numbers
#
# the numbers are poped from the stack
# and the answer is pushed onto the stack
sub ADD {
	my $self = shift;

	my $b = $self->POP;
	my $a = $self->POP;

	$self->PUSH( $a + $b );

	return;
}

# substracts two numbers
#
# the numbers are poped from the stack
# and the answer is pushed onto the stack
sub SUBST {
	my $self = shift;

	my $b = $self->POP;
	my $a = $self->POP;

	$self->PUSH( $a - $b );

	return;
}

# multiplies two numbers
#
# the numbers are poped from the stack
# and the answer is pushed onto the stack
sub MUL {
	my $self = shift;

	my $b = $self->POP;
	my $a = $self->POP;

	$self->PUSH( $a * $b );

	return;
}

# divides two numbers
#
# the numbers are poped from the stack
# and the answer is pushed onto the stack
sub DIV {
	my $self = shift;

	my $b = $self->POP;
	my $a = $self->POP;

	$self->PUSH( $a / $b );

	return;
}

# gets the modulus of two numbers
#
# the numbers are poped from the stack
# and the answer is pushed onto the stack
sub MOD {
	my $self = shift;

	my $b = $self->POP;
	my $a = $self->POP;

	$self->PUSH( $a % $b );

	return;
}

# BITWISE

# shift the a number left N number of possitions
#
# the numbers are poped from the stack
# and the answer is pushed onto the stack
sub SHL {
	my $self = shift;

	my $b = $self->POP;
	my $a = $self->POP;

	$self->PUSH( $a << $b );

	return;
}

# shift the a number right N number of possitions
#
# the numbers are poped from the stack
# and the answer is pushed onto the stack
sub SHR {
	my $self = shift;

	my $b = $self->POP;
	my $a = $self->POP;

	$self->PUSH( $a >> $b );

	return;
}

# ANDs two numbers
#
# the numbers are poped from the stack
# and the answer is pushed onto the stack
sub AND {
	my $self = shift;

	my $b = $self->POP;
	my $a = $self->POP;

	$self->PUSH( $a & $b );

	return;
}

# ORs two numbers
#
# the numbers are poped from the stack
# and the answer is pushed onto the stack
sub OR {
	my $self = shift;

	my $b = $self->POP;
	my $a = $self->POP;

	$self->PUSH( $a | $b );

	return;
}

# XORs two numbers
#
# the numbers are poped from the stack
# and the answer is pushed onto the stack
sub XOR {
	my $self = shift;

	my $b = $self->POP;
	my $a = $self->POP;

	$self->PUSH( $a ^ $b );

	return;
}

# complement of a number
#
# the numbers are poped from the stack
# and the answer is pushed onto the stack
sub COMP {
	my $self = shift;

	my $a = $self->POP;

	$self->PUSH( ~$a );

	return;
}

# BRANCHING
sub JMP {
	my $self = shift;

	# get the instruction to jump to
	return $self->ip( $self->POP );
}

sub JMPEQ {
	my $self = shift;

	my $jmp = $self->POP;    # pop the jump location

	my $b = $self->POP;
	my $a = $self->POP;

	my $res;
	if ( $self->_is_numeric_pair( $a, $b ) ) {
		$res = $a == $b;
	} else {
		$res = $a eq $b;
	}

	if ($res) {
		$self->PUSH($jmp);
		$self->JMP;
	}

	return;
}

sub JMPNEQ {
	my $self = shift;

	my $jmp = $self->POP;    # pop the jump location

	my $b = $self->POP;
	my $a = $self->POP;

	my $res;
	if ( $self->_is_numeric_pair( $a, $b ) ) {
		$res = $a != $b;
	} else {
		$res = $a ne $b;
	}

	if ($res) {
		$self->PUSH($jmp);
		$self->JMP;
	}

	return;
}

sub JMPLT {
	my $self = shift;

	my $jmp = $self->POP;    # pop the jump location

	my $b = $self->POP;
	my $a = $self->POP;

	my $res;
	if ( $self->_is_numeric_pair( $a, $b ) ) {
		$res = $a < $b;
	} else {
		$res = $a lt $b;
	}

	if ( $a < $b ) {
		$self->PUSH($jmp);
		$self->JMP;
	}

	return;
}

sub JMPGT {
	my $self = shift;

	my $jmp = $self->POP;    # pop the jump location

	my $b = $self->POP;
	my $a = $self->POP;

	my $res;
	if ( $self->_is_numeric_pair( $a, $b ) ) {
		$res = $a > $b;
	} else {
		$res = $a gt $b;
	}

	if ($res) {
		$self->PUSH($jmp);
		$self->JMP;
	}

	return;
}

# SUBRUTINE CALLS

sub CALL {
	my $self = shift;

	push @{ $self->address_stack },
	  $self->ip;    # push the return location onto the call stack

	$self->JMP;     # jump to the call

	return;
}

sub RET {
	my $self = shift;

	$self->PUSH( pop @{ $self->address_stack } )
	  ;             # pop the return location from the call stack
	$self->JMP;     # return to calle

	return;
}

# PSEUDO I/O HANDLERS

sub OUT {
	my $self = shift;

	my $data = $self->POP;    # pop the data into an internal register

	#print the text
	if ( !defined fileno($data) ) {
		if ( ref $data =~ /SCALAR/ ) {
			print $$data;
		} elsif ( ref $data !~ /SCALAR/ ) {
			$self->_trace("Operation not suported!\n");
		} else {
			print $data;
		}
	} else {

		# print the file handle out line by line
		while (<$data>) {
			print;
		}
	}

	return;
}

sub IN {
	my $self = shift;

	my ( $addr, $size ) = $self->_find_free_block(1);

	my $fh   = $self->POP;    # pop the filehandle
	my $data = <$fh>;         # read the data in

	$self->PUSH($data);       #push the data onto the heap
	$self->PUSH($addr);
	$self->STORE;
	$self->PUSH($addr);       #push the address again so that

	return;
}

# HEAP HANDLERS

sub LOAD {
	my $self = shift;

	my $addr = $self->POP;

	$self->PUSH( $self->heap->[$addr]->[1] );

	return;
}

sub STORE {
	my $self = shift;

	my $addr = $self->POP;
	my $val  = $self->_peek; # don't remove from stack

	$self->_trace("Address out of bounds!") if $addr > 255;

	$self->heap->[$addr]->[0] = 1;
	$self->heap->[$addr]->[1] = $val;

	return;
}

# clears an anddress
sub CLS {
	my $self = shift;

	my $addr = $self->POP;

	$self->heap->[$addr]->[0] = 0;
	$self->heap->[$addr]->[1] = undef;

	return;
}

no Moose;
__PACKAGE__->meta->make_immutable;

1;

=pod

=head1 SUPPORT

No support is available

=head1 AUTHOR

Copyright 2011 Anonymous.

=cut
